greenpithumb/poller.py

```diff
@@ -73,12 +97,54 @@ def __init__(self, local_clock, poll_interval, record_queue, sensor):
         self._sensor = sensor
         self._stopped = threading.Event()
 
+    def _unix_now(self):
+        return _datetime_to_unix_time(self._local_clock.now())
+
+    def _wait_until_unix_time(self, target_unix_time):
+        """Waits until the target time arrives or the worker is stopped.
+
+        Args:
+            target_unix_time: UNIX time to wait until.
+        """
+        logger.info('%s is waiting %d seconds until next poll',
+                    self.__class__.__name__,
+                    target_unix_time - self._unix_now())
+        while not self._is_stopped() and (self._unix_now() < target_unix_time):
+            self._local_clock.wait(_IDLE_SECONDS)
+
+    def _next_poll_time(self, last_poll_time):
+        """Calculates time of next poll in UNIX time.
+
+        Calculates time of next poll so that it is a multiple of
+        self._poll_interval. If the next multiple is the same as the last poll
+        time, returns a poll time that is the current time + one poll interval.
+
+        Args:
+            last_poll_time: The UNIX time of the last poll or None if this is
+                the first poll.
+
+        Returns:
+            UNIX time of next scheduled poll.
+        """
+        next_poll_time = _round_up_to_multiple(self._unix_now(),
+                                               self._poll_interval)
+        if next_poll_time and (next_poll_time == last_poll_time):
```

Comment:
>I'm confused here. next_poll_time will never be False, right?

Oh, whoops, that was supposed to be `last_poll_time`. Fixed.

>Also next_poll_time and last_poll_time will always be different since total_seconds() returns to microsecond precision?

Oh yeah, you're right. I didn't realize `total_seconds` was a float instead of an int. It *theoretically* could end up the same value if it managed to poll in less than 1 microsecond (which probably can't happen practically), but I actually meant for the UNIX time to be integer seconds because I feel like it's kind of confusing / unexpected to see mod operations on a float. So I've updated `_unix_now` to return an int and now the check is more necessary because there can be a time collision if the sensor reading takes less than one second.